File: apps/employee_app/lib/data/roster_assignments.dart
/// Mirrors `apps/admin-web/src/data/rosterAssignments.ts`.

import 'customers_data.dart';
import 'staff_directory.dart';

final class DayPosting {
  const DayPosting({
    required this.staffSlug,
    required this.staffName,
    required this.customerSlug,
    required this.customerName,
    required this.siteName,
  });

  final String staffSlug;
  final String staffName;
  final String customerSlug;
  final String customerName;
  final String siteName;
}

final class SiteColleagueGroup {
  const SiteColleagueGroup({
    required this.customerSlug,
    required this.customerName,
    required this.siteName,
    required this.colleagues,
  });

  final String customerSlug;
  final String customerName;
  final String siteName;
  final List<({String slug, String name})> colleagues;
}

int _hashString(String s) {
  var h = 0;
  for (var i = 0; i < s.length; i++) {
    h = (31 * h + s.codeUnitAt(i)) & 0x7fffffff;
  }
  return h;
}

List<({String customerSlug, String customerName, String siteName})>
    eligiblePostingsForStaffName(String name) {
  final out = <({String customerSlug, String customerName, String siteName})>[];
  for (final c in kCustomerRecords) {
    for (final site in c.sites) {
      if (site.guards.contains(name)) {
        out.add((
          customerSlug: c.slug,
          customerName: c.name,
          siteName: site.siteName,
        ));
      }
    }
  }
  return out;
}

DayPosting postingForDay(String staffSlug, String staffName, String dateKey) {
  final eligible = eligiblePostingsForStaffName(staffName);
  if (eligible.isEmpty) {
    return DayPosting(
      staffSlug: staffSlug,
      staffName: staffName,
      customerSlug: 'unassigned',
      customerName: 'Unassigned',
      siteName: '—',
    );
  }
  final i = _hashString(dateKey + staffSlug) % eligible.length;
  final p = eligible[i];
  return DayPosting(
    staffSlug: staffSlug,
    staffName: staffName,
    customerSlug: p.customerSlug,
    customerName: p.customerName,
    siteName: p.siteName,
  );
}

DayPosting postingForShift(
  String staffSlug,
  String staffName,
  String dateKey,
  int shiftIndex,
) {
  final eligible = eligiblePostingsForStaffName(staffName);
  if (eligible.isEmpty) {
    return DayPosting(
      staffSlug: staffSlug,
      staffName: staffName,
      customerSlug: 'unassigned',
      customerName: 'Unassigned',
      siteName: '—',
    );
  }
  final i =
      _hashString('$dateKey:$staffSlug:shift$shiftIndex') % eligible.length;
  final p = eligible[i];
  return DayPosting(
    staffSlug: staffSlug,
    staffName: staffName,
    customerSlug: p.customerSlug,
    customerName: p.customerName,
    siteName: p.siteName,
  );
}

const List<String> kScheduleShiftLabels = [
  '00:00–08:00',
  '08:00–16:00',
  '16:00–00:00',
];

sealed class PersonalScheduleSegment {
  const PersonalScheduleSegment({required this.label});

  final String label;
}

final class OffSegment extends PersonalScheduleSegment {
  const OffSegment({required super.label});
}

final class WorkSegment extends PersonalScheduleSegment {
  const WorkSegment({required super.label, required this.posting});

  final DayPosting posting;
}

/// Five shifts per week: one day shift Mon–Fri; weekends & other bands off.
List<PersonalScheduleSegment> personalScheduleThreeSegments(
  String staffSlug,
  String staffName,
  String dateKey,
) {
  final parts = dateKey.split('-');
  final y = int.parse(parts[0]);
  final m = int.parse(parts[1]);
  final d = int.parse(parts[2]);
  final date = DateTime(y, m, d);
  final dow = date.weekday; // Mon=1 .. Sun=7
  final isWeekend = dow == DateTime.saturday || dow == DateTime.sunday;

  if (isWeekend) {
    return [
      OffSegment(label: kScheduleShiftLabels[0]),
      OffSegment(label: kScheduleShiftLabels[1]),
      OffSegment(label: kScheduleShiftLabels[2]),
    ];
  }

  final dayShift = postingForShift(staffSlug, staffName, dateKey, 1);
  return [
    OffSegment(label: kScheduleShiftLabels[0]),
    WorkSegment(label: kScheduleShiftLabels[1], posting: dayShift),
    OffSegment(label: kScheduleShiftLabels[2]),
  ];
}

List<DayPosting> buildAssignmentsForDay(String dateKey) {
  return kStaffRecords
      .map((s) => postingForDay(s.slug, s.name, dateKey))
      .toList(growable: false);
}

List<DayPosting> filterPostingsForClient(
  List<DayPosting> postings,
  String? clientSlug,
) {
  if (clientSlug == null) return postings;
  return postings.where((p) => p.customerSlug == clientSlug).toList();
}

List<String> staffSlugsForCustomer(String customerSlug) {
  final c = customerBySlug(customerSlug);
  if (c == null) return [];
  final names = <String>{};
  for (final site in c.sites) {
    names.addAll(site.guards);
  }
  final slugs = <String>[];
  for (final name in names) {
    final rec = staffByName(name);
    if (rec != null) slugs.add(rec.slug);
  }
  return slugs;
}

List<SiteColleagueGroup> sitesAndColleaguesForStaff(String staffName) {
  final groups = <SiteColleagueGroup>[];
  for (final c in kCustomerRecords) {
    for (final site in c.sites) {
      if (!site.guards.contains(staffName)) continue;
      final colleagues = <({String slug, String name})>[];
      for (final g in site.guards) {
        if (g == staffName) continue;
        final rec = staffByName(g);
        if (rec != null) colleagues.add((slug: rec.slug, name: rec.name));
      }
      groups.add(
        SiteColleagueGroup(
          customerSlug: c.slug,
          customerName: c.name,
          siteName: site.siteName,
          colleagues: colleagues,
        ),
      );
    }
  }
  return groups;
}

String dateKeyFromParts(int year, int monthIndex, int day) {
  final m = (monthIndex + 1).toString().padLeft(2, '0');
  final d = day.toString().padLeft(2, '0');
  return '$year-$m-$d';
}

File: apps/admin-web/src/data/rosterAssignments.ts
import { CUSTOMER_RECORDS, getCustomerBySlug } from "./customers";
import { getStaffByName, STAFF_RECORDS } from "./staffDirectory";

export type SiteColleagueGroup = {
  customerSlug: string;
  customerName: string;
  siteName: string;
  colleagues: Array<{ slug: string; name: string }>;
};

/** Sites this officer covers, with other guards on the same site (shift overlap / handover). */
export function getSitesAndColleaguesForStaff(staffName: string): SiteColleagueGroup[] {
  const groups: SiteColleagueGroup[] = [];
  for (const c of CUSTOMER_RECORDS) {
    for (const site of c.sites) {
      if (!site.guards.includes(staffName)) continue;
      const colleagues = site.guards
        .filter((g) => g !== staffName)
        .map((name) => {
          const rec = getStaffByName(name);
          return rec ? { slug: rec.slug, name: rec.name } : null;
        })
        .filter((x): x is { slug: string; name: string } => x != null);
      groups.push({
        customerSlug: c.slug,
        customerName: c.name,
        siteName: site.siteName,
        colleagues,
      });
    }
  }
  return groups;
}

export type DayPosting = {
  staffSlug: string;
  staffName: string;
  customerSlug: string;
  customerName: string;
  siteName: string;
};

function hashString(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) h = (Math.imul(31, h) + s.charCodeAt(i)) | 0;
  return Math.abs(h);
}

export function getEligiblePostingsForStaffName(name: string): Array<{
  customerSlug: string;
  customerName: string;
  siteName: string;
}> {
  const out: Array<{ customerSlug: string; customerName: string; siteName: string }> = [];
  for (const c of CUSTOMER_RECORDS) {
    for (const site of c.sites) {
      if (site.guards.includes(name)) {
        out.push({ customerSlug: c.slug, customerName: c.name, siteName: site.siteName });
      }
    }
  }
  return out;
}

/** Three roster blocks per calendar day (personal schedule). */
export const SCHEDULE_SHIFT_LABELS = ["00:00–08:00", "08:00–16:00", "16:00–00:00"] as const;

export function postingForDay(staffSlug: string, staffName: string, dateKey: string): DayPosting {
  const eligible = getEligiblePostingsForStaffName(staffName);
  if (eligible.length === 0) {
    return {
      staffSlug,
      staffName,
      customerSlug: "unassigned",
      customerName: "Unassigned",
      siteName: "—",
    };
  }
  const i = hashString(dateKey + staffSlug) % eligible.length;
  const p = eligible[i];
  return { staffSlug, staffName, ...p };
}

/** One posting per 8-hour block; index 0 = night, 1 = day, 2 = evening. */
export function postingForShift(
  staffSlug: string,
  staffName: string,
  dateKey: string,
  shiftIndex: 0 | 1 | 2,
): DayPosting {
  const eligible = getEligiblePostingsForStaffName(staffName);
  if (eligible.length === 0) {
    return {
      staffSlug,
      staffName,
      customerSlug: "unassigned",
      customerName: "Unassigned",
      siteName: "—",
    };
  }
  const i = hashString(`${dateKey}:${staffSlug}:shift${shiftIndex}`) % eligible.length;
  const p = eligible[i];
  return { staffSlug, staffName, ...p };
}

function parseLocalDateKey(dateKey: string): Date {
  const [y, m, d] = dateKey.split("-").map(Number);
  return new Date(y, m - 1, d);
}

/** One row in the personal schedule grid (three per day). */
export type PersonalScheduleSegment =
  | { label: string; status: "off" }
  | { label: string; status: "work"; posting: DayPosting };

/**
 * Personal schedule: **5 shifts per week** — one day shift (08:00–16:00) Mon–Fri only.
 * Nights, evenings, and all weekend segments are off (no 24/7 coverage in mock data).
 */
export function personalScheduleThreeSegments(
  staffSlug: string,
  staffName: string,
  dateKey: string,
): PersonalScheduleSegment[] {
  const labels = SCHEDULE_SHIFT_LABELS;
  const d = parseLocalDateKey(dateKey);
  const dow = d.getDay();
  const isWeekend = dow === 0 || dow === 6;

  if (isWeekend) {
    return [
      { label: labels[0], status: "off" },
      { label: labels[1], status: "off" },
      { label: labels[2], status: "off" },
    ];
  }

  const dayShift = postingForShift(staffSlug, staffName, dateKey, 1);
  return [
    { label: labels[0], status: "off" },
    { label: labels[1], status: "work", posting: dayShift },
    { label: labels[2], status: "off" },
  ];
}

export function buildAssignmentsForDay(dateKey: string): DayPosting[] {
  return STAFF_RECORDS.map((s) => postingForDay(s.slug, s.name, dateKey));
}

export function dateKeyFromParts(year: number, monthIndex: number, day: number): string {
  const m = String(monthIndex + 1).padStart(2, "0");
  const d = String(day).padStart(2, "0");
  return `${year}-${m}-${d}`;
}

/** Staff slugs that guard any site under this customer (for map + filters). */
export function getStaffSlugsForCustomer(customerSlug: string): string[] {
  const c = getCustomerBySlug(customerSlug);
  if (!c) return [];
  const names = new Set<string>();
  for (const s of c.sites) {
    for (const g of s.guards) names.add(g);
  }
  const slugs: string[] = [];
  for (const name of names) {
    const rec = getStaffByName(name);
    if (rec) slugs.push(rec.slug);
  }
  return slugs;
}

export function filterPostingsForClient(postings: DayPosting[], clientSlug: string | null): DayPosting[] {
  if (!clientSlug) return postings;
  return postings.filter((p) => p.customerSlug === clientSlug);
}
